// n8n-nodes-tmdb/dist/credentials/TmdbApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TmdbApi = void 0;
class TmdbApi {
    constructor() {
        this.name = 'tmdbApi';
        this.displayName = 'TMDB API';
        this.iconUrl = 'file:tmdb.svg';
        this.documentationUrl = 'https://developer.themoviedb.org/reference/intro/getting-started';
        this.properties = [
            {
                displayName: 'Api Token',
                name: 'api_token',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '={{"Bearer " + $credentials.api_token}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.themoviedb.org/3',
                url: '/configuration',
            },
        };
    }
}
exports.TmdbApi = TmdbApi;
