// n8n-nodes-tmdb/dist/nodes/TmdbNode/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPosterUrls = getPosterUrls;
exports.getLogoUrls = getLogoUrls;
exports.getBackdropUrls = getBackdropUrls;
exports.getProfileUrls = getProfileUrls;
exports.getStillUrls = getStillUrls;
function getUrls(filePath, sizes) {
    if (filePath === null) {
        return null;
    }
    const baseUrl = 'https://image.tmdb.org/t/p/';
    const urls = {};
    sizes.forEach((size) => {
        urls[size] = `${baseUrl}${size}/${filePath}`;
    });
    return urls;
}
function getPosterUrls(filePath) {
    const posterSizes = ['original', 'w92', 'w154', 'w185', 'w342', 'w500', 'w780'];
    return getUrls(filePath, posterSizes);
}
function getLogoUrls(filePath) {
    const logoSizes = ['original', 'w45', 'w92', 'w154', 'w185', 'w300', 'w500'];
    return getUrls(filePath, logoSizes);
}
function getBackdropUrls(filePath) {
    const backdropSizes = ['original', 'w300', 'w780', 'w1280'];
    return getUrls(filePath, backdropSizes);
}
function getProfileUrls(filePath) {
    const profileSizes = ['original', 'w45', 'w185', 'h632'];
    return getUrls(filePath, profileSizes);
}
function getStillUrls(filePath) {
    const stillSizes = ['original', 'w92', 'w185', 'w300'];
    return getUrls(filePath, stillSizes);
}
