// n8n-nodes-tmdb/dist/nodes/TmdbNode/MovieListFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMovieList = getMovieList;
const UrlFunctions_1 = require("./UrlFunctions");
async function getMovieList(context, itemIndex) {
    const language = context.getNodeParameter('language', itemIndex);
    const movieListType = context.getNodeParameter('movie_list_type', itemIndex);
    const page = context.getNodeParameter('page', itemIndex);
    const region = context.getNodeParameter('region', itemIndex);
    const response = await context.helpers.httpRequestWithAuthentication.call(context, 'tmdbApi', {
        headers: {
            Accept: 'application/json',
        },
        method: 'GET',
        url: `https://api.themoviedb.org/3/movie/${movieListType}?language=${language}&page=${page}&region=${region}`,
    });
    if (response.results) {
        response.results = response.results.map((movie) => ({
            ...movie,
            backdrop_urls: (0, UrlFunctions_1.getBackdropUrls)(movie.backdrop_path),
            poster_urls: (0, UrlFunctions_1.getBackdropUrls)(movie.poster_path),
        }));
    }
    return response;
}
