// n8n-nodes-tmdb/dist/nodes/TmdbNode/MovieFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMovie = getMovie;
const UrlFunctions_1 = require("./UrlFunctions");
async function getMovie(context, itemIndex) {
    var _a, _b, _c, _d, _e, _f;
    const language = context.getNodeParameter('language', itemIndex);
    const movieId = context.getNodeParameter('movie_id', itemIndex);
    const appendToResponse = context.getNodeParameter('movie_append_to_response', itemIndex);
    const response = await context.helpers.httpRequestWithAuthentication.call(context, 'tmdbApi', {
        headers: {
            Accept: 'application/json',
        },
        method: 'GET',
        url: `https://api.themoviedb.org/3/movie/${movieId}?language=${language}&append_to_response=${appendToResponse}`,
    });
    if (response.belongs_to_collection) {
        response.belongs_to_collection = {
            ...response.belongs_to_collection,
            poster_urls: (0, UrlFunctions_1.getPosterUrls)(response.belongs_to_collection.poster_path),
            backdrop_urls: (0, UrlFunctions_1.getBackdropUrls)(response.belongs_to_collection.backdrop_path),
        };
    }
    if (response.production_companies) {
        response.production_companies = response.production_companies.map((company) => ({
            ...company,
            logo_urls: (0, UrlFunctions_1.getLogoUrls)(company.logo_path),
        }));
    }
    if (response.images) {
        response.images.backdrops = (_b = (_a = response.images) === null || _a === void 0 ? void 0 : _a.backdrops) === null || _b === void 0 ? void 0 : _b.map((backdrop) => ({
            ...backdrop,
            urls: (0, UrlFunctions_1.getBackdropUrls)(backdrop.file_path),
        }));
        response.images.logos = (_d = (_c = response.images) === null || _c === void 0 ? void 0 : _c.logos) === null || _d === void 0 ? void 0 : _d.map((logo) => ({
            ...logo,
            urls: (0, UrlFunctions_1.getLogoUrls)(logo.file_path),
        }));
        response.images.posters = (_f = (_e = response.images) === null || _e === void 0 ? void 0 : _e.posters) === null || _f === void 0 ? void 0 : _f.map((poster) => ({
            ...poster,
            urls: (0, UrlFunctions_1.getPosterUrls)(poster.file_path),
        }));
    }
    if (response.credits) {
        response.credits.cast = response.credits.cast.map((cast) => ({
            ...cast,
            profile_urls: (0, UrlFunctions_1.getProfileUrls)(cast.profile_path),
        }));
        response.credits.crew = response.credits.crew.map((crew) => ({
            ...crew,
            profile_urls: (0, UrlFunctions_1.getProfileUrls)(crew.profile_path),
        }));
    }
    return {
        ...response,
        poster_urls: (0, UrlFunctions_1.getPosterUrls)(response.poster_path),
        backdrop_urls: (0, UrlFunctions_1.getBackdropUrls)(response.backdrop_path),
    };
}
